include/glow/Graph/Log.h

```diff
@@ -0,0 +1,100 @@
+/**
+ * Copyright (c) 2018-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef GLOW_SUPPORT_LOG_H
+#define GLOW_SUPPORT_LOG_H
+
+#include "llvm/ADT/StringRef.h"
+#include "llvm/ADT/ilist.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include <fstream>
+#include <string>
+
+namespace glow {
+
+class Node;
+
+extern llvm::cl::opt<bool> dumpCompilationLogOpt;
+
+/// A class for logging all compilation related activities.
+class LogContext {
+private:
+  /// A vector that keeps track of all log contents.
+  std::vector<std::string> logContents_;
+
+  /// A vector that keeps track of current compilation scopes.
+  std::vector<std::string> logScopes_;
+
+  /// A string that represents the current full scope name.
+  std::string currentFullScope_;
+
+public:
+  /// Add content into the contents vector.
+  void addLogContent(llvm::StringRef logContent);
+
+  /// Add a scope into the scope vector.
+  void pushLogScope(llvm::StringRef scopeName);
+
+  /// Getter that returns the current full scope name.
+  llvm::StringRef getFullScopeName() { return currentFullScope_; }
+
+  /// Pops out the most recently added scope.
+  void popLogScope();
+
+  /// Dumps the log into a file named after the given \p funcName.
+  void dumpLog(llvm::StringRef funcName);
+
+  /// Logs the node creation with a list of input nodes.
+  void logNodeCreation(const Node *newNode);
+
+  /// Logs the node replacement.
+  void logNodeReplacement(const Node *oldNode, const Node *newNode);
+
+  /// Logs the node deletion.
+  void logNodeDeletion(const Node &deletedNode);
+};
+
+/// Logs a new log scope.
+#define LOG_SCOPE(ctx, name) ScopedLogBlock __scope__(ctx, name);
+
+/// Helper class which traces the start and end of a compilation log scope.
+class ScopedLogBlock {
+
+  /// Reference to the log context.
+  LogContext &ctx_;
+
+  /// The name of the log scope.
+  llvm::StringRef name_;
```

Comment:
Is it a good idea that it is a StringRef? If names are always constant string literals, it would work. But if the name is ever created dynamically it may result in some issues I'm afraid.

Reply:
make sense